### DISCUSSION TITLE
Relationship load options affect following queries

### DISCUSSION THREAD

**Discussion by mattjudge** [sqlalchemy/sqlalchemy#11040] [Usage Questions]
### Describe the bug

1. Select a row from table A with a subqueryload joining table B, and noload option on that subquery B's relationship with A
2. The result is returned as expected, including table A's relationship with table B.
3. Select a row from table B with a subqueryload joining table A, and noload option on that subquery A's relationship with B
4. The result does not include table B's relationship with table A. If step 1 is omitted, it is included.

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html

### SQLAlchemy Version in Use

1.4.50

### DBAPI (i.e. the database driver)

pysqlite, psycopg2

### Database Vendor and Major Version

SQLite, PostgreSQL

### Python Version

3.11.6

### Operating system

WSL

### To Reproduce

```python
from sqlalchemy import Column, ForeignKey, Integer, create_engine, insert, select
from sqlalchemy.orm import Session, declarative_base, noload, relationship, subqueryload

Base = declarative_base()


# two tables, with back_populates
class A(Base):
    __tablename__ = "a"

    a_id = Column(
        Integer,
        nullable=False,
        primary_key=True,
        autoincrement=True,
    )
    b_connections = relationship("B", back_populates="a_connections", primaryjoin="foreign(B.a_id) == A.a_id", uselist=True)


class B(Base):
    __tablename__ = "b"

    b_id = Column(
        Integer,
        nullable=False,
        primary_key=True,
        autoincrement=True,
    )
    a_id = Column(
        ForeignKey("a.a_id"),
        nullable=False,
    )
    a_connections = relationship("A", back_populates="b_connections", primaryjoin="A.a_id == foreign(B.a_id)", uselist=True)


engine = create_engine("sqlite://", echo=False)
Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

# initial setup block
with Session(engine) as session:
    a_insert = session.execute(insert(A).values())
    a_id = a_insert.lastrowid
    b_insert = session.execute(insert(B).values(a_id=a_id))
    b_id = b_insert.lastrowid

    session.commit()


a_stmt = (
    select(A)
    .options(
        subqueryload(A.b_connections).options(noload(B.a_connections)),
    )
    .where(A.a_id == a_id)
)
b_stmt = (
    select(B)
    .options(
        subqueryload(B.a_connections).options(noload(A.b_connections)),
    )
    .where(B.b_id == b_id)
)

with Session(engine) as session:
    # these blocks can be run in either order, the second one always fails
    (a, *_), *_ = session.execute(a_stmt).fetchall()
    assert a_id == a.a_id
    assert 1 == len(a.b_connections)
    a_b = a.b_connections[0]
    assert b_id == a_b.b_id

    (b, *_), *_ = session.execute(b_stmt).fetchall()
    assert b_id == b.b_id
    assert 1 == len(b.a_connections)  # this assertion fails
    b_a = b.a_connections[0]
    assert a_id == b_a.a_id
```

### Error

```
Traceback (most recent call last):
  File "...", line 73, in <module>
    assert 1 == len(b.a_connections)  # this assertion fails
           ^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

**Comment by CaselIT** [comment]
Hi,

The session will deduplicate the incoming objects with the ones that are already in the session, so the you get on the second query is the same b you obtained from the fist query, meaning that this assertion should pass `assert b is a_b`

You can use populate existing to update the data in the session (not sure if it works in this case, but it should) https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#populate-existing

**Comment by mattjudge** [comment]
Thanks for the quick reply! Indeed, `b is a_b`. I would have expected the load options to be taken into account when deduplicating the incoming objects, since otherwise every select needs to be aware of all other selects and patterns used within the same session? e.g. using populate_existing does indeed pass my previous tests, but would fail the following test, because it overwrites the first object with the second unloaded object.

```
    (a, *_), *_ = session.execute(a_stmt).fetchall()
    assert a_id == a.a_id
    assert 1 == len(a.b_connections)
    a_b = a.b_connections[0]
    assert b_id == a_b.b_id

    (b, *_), *_ = session.execute(b_stmt).fetchall()
    assert b_id == b.b_id
    assert 1 == len(b.a_connections)
    b_a = b.a_connections[0]
    assert a_id == b_a.a_id

    assert a_id == a.a_id
    assert 1 == len(a.b_connections)  # this assertion fails
    assert b_id == a_b.b_id

```

 Is there an alternative pattern we should be using for this type of data load?

**Reply by CaselIT** [reply]
I don't think thats currently supported, but I may be wrong. @zzzeek is there a way to achieve that?

**Reply by zzzeek** [reply]
if you want to load objects and have everything set up fresh, you use populate existing.  is that not working?

**Reply by CaselIT** [reply]
it is but it seems they are also setting a noload in the second query and that's removing loaded existing things.

I guess the best option would be to remove `.options(noload(A.b_connections)),` from the second query

**Reply by zzzeek** [reply]
oh.   dont use noload.   @CaselIT I think we should deprecate noload since it leads to poor results and there is no valid reason to use it

**Reply by zzzeek** [reply]
added #11045

**Reply by mattjudge** [reply]
You're correct in #11045 that this originally stemmed from our use of a custom serializer. The fundamental problem is that the mutual back_populates results in an infinite reference loop, unless we use noload, due to the nature above of `b is a_b`. Is this a problem that other people have overcome before? Are you aware of any existing patterns for this type of problem?

**Reply by mattjudge** [reply]
(is raiseload supported, or similarly problematic?)

**Reply by CaselIT** [reply]
> back_populates results in an infinite reference loop

this is a limitation of the serializer you are using that does not support loop.
Either change it to support loops or allow it to skip serializing some properties

**Reply by zzzeek** [reply]
raiseload is great because it doesnt change the state of the mapped object.   it's just noload that has all these dead-end cases,  I had tried to improve them a couple of years ago but overall it's a loader that returns the wrong information.

anyway yeah serializers by definition have to keep a set of objects that have already been serialized and avoid recursion loops.
